h2o-automl/src/main/java/ai/h2o/automl/TrainingStepsRegistry.java

```diff
@@ -0,0 +1,106 @@
+package ai.h2o.automl;
+
+import ai.h2o.automl.EventLogEntry.Stage;
+import ai.h2o.automl.StepDefinition.Alias;
+import ai.h2o.automl.StepDefinition.Step;
+import water.Iced;
+import water.nbhm.NonBlockingHashMap;
+import water.util.ArrayUtils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The registry responsible for loading all {@link TrainingStepsProvider} using service discovery,
+ * and providing the list of {@link TrainingStep} to execute.
+ */
+public class TrainingStepsRegistry extends Iced<TrainingStepsRegistry> {
+
+    static final NonBlockingHashMap<String, Class<TrainingSteps>> stepsByName = new NonBlockingHashMap<>();
+
+    static {
+        ServiceLoader<TrainingStepsProvider> trainingStepsProviders = ServiceLoader.load(TrainingStepsProvider.class);
+        for (TrainingStepsProvider provider : trainingStepsProviders) {
+            stepsByName.put(provider.getName(), provider.getStepsClass());
+        }
+    }
+
+    private StepDefinition[] _defaultTrainingPlan;
+
+    public TrainingStepsRegistry(StepDefinition[] defaultTrainingPlan) {
+        _defaultTrainingPlan = defaultTrainingPlan;
+    }
+
+    /**
+     * @param aml the AutoML instance responsible to execute the {@link TrainingStep}s.
+     * @return the list of {@link TrainingStep}s to execute according to the default training plan.
+     */
+    public TrainingStep[] getOrderedSteps(AutoML aml) {
+        return getOrderedSteps(aml, _defaultTrainingPlan);
+    }
+
+
+    /**
+     * @param aml the AutoML instance responsible to execute the {@link TrainingStep}s.
+     * @return the list of {@link TrainingStep}s to execute according to the given training plan.
+     */
+    public TrainingStep[] getOrderedSteps(AutoML aml, StepDefinition[] trainingPlan) {
+        aml.eventLog().info(Stage.Workflow, "Loading execution steps: "+Arrays.toString(trainingPlan));
+        List<TrainingStep> orderedSteps = new ArrayList<>();
+        for (StepDefinition def : trainingPlan) {
+            Class<TrainingSteps> clazz = stepsByName.get(def._name);
```

Comment:
imho this could be simplified if the providers were itself the factory classes

Reply:
agree